File: 백준/Silver/16401. 과자 나눠주기/과자 나눠주기.c
#include<stdio.h>
#define int long long

int n,m;
int arr[1010101];

int check(int x){
	int ret = 0;
	for (int i = 0; i < m; i++){
		ret += arr[i]/x;
	}
	return ret >= n;
}

signed main(){
	scanf("%lld %lld", &n, &m);
	for (int i = 0; i < m; i++) scanf("%lld", &arr[i]);
	
	int left = 0;
	int right = 1e18;
	
	while (left + 1 < right){
		int mid = (left+right)/2;
		
		if (check(mid)) left = mid;
		else right = mid;
	}
	
	printf("%lld", left);
	return 0;
}

File: 백준/Silver/16401. 과자 나눠주기/과자 나눠주기.py
import sys
input = sys.stdin.readline

def check(x):
    ret = 0
    for i in range(m):
        ret += l[i]//x
    return ret >= n

n,m = map(int, input().split())
l = list(map(int, input().split()))

left = 0
right = int(1e20)

while left + 1 < right:
    mid = (left+right)//2
    
    if check(mid): left = mid
    else: right = mid

print(left)

